File: typescript/day11_part2_2022.ts
import * as fs from 'fs';

interface Monkey {
    items: number[];
    operation: (old: number) => number;
    test: number;
    trueMonkey: number;
    falseMonkey: number;
    inspectedCount: number;
}

function parseInput(input: string): Monkey[] {
    const monkeys: Monkey[] = [];
    const lines = input.trim().split('\n\n');

    for (const block of lines) {
        const lines = block.split('\n');
        const items = lines[1].match(/\d+/g)!.map(Number);
        const operationLine = lines[2].split('=')[1].trim();
        const operation = new Function('old', `return ${operationLine}`) as (old: number) => number;
        const test = Number(lines[3].match(/\d+/)![0]);
        const trueMonkey = Number(lines[4].match(/\d+/)![0]);
        const falseMonkey = Number(lines[5].match(/\d+/)![0]);

        monkeys.push({ items, operation, test, trueMonkey, falseMonkey, inspectedCount: 0 });
    }

    return monkeys;
}

function simulateMonkeys(monkeys: Monkey[], rounds: number) {
    const modulo = monkeys.reduce((acc, monkey) => acc * monkey.test, 1);

    for (let round = 0; round < rounds; round++) {
        for (const monkey of monkeys) {
            while (monkey.items.length > 0) {
                const item = monkey.items.shift()!;
                monkey.inspectedCount++;
                const newWorryLevel = monkey.operation(item) % modulo;

                if (newWorryLevel % monkey.test === 0) {
                    monkeys[monkey.trueMonkey].items.push(newWorryLevel);
                } else {
                    monkeys[monkey.falseMonkey].items.push(newWorryLevel);
                }
            }
        }
    }
}

function calculateMonkeyBusiness(monkeys: Monkey[]): number {
    const inspectionCounts = monkeys.map(m => m.inspectedCount);
    inspectionCounts.sort((a, b) => b - a);
    return inspectionCounts[0] * inspectionCounts[1];
}

function main() {
    const input = fs.readFileSync('input.txt', 'utf-8');
    const monkeys = parseInput(input);
    simulateMonkeys(monkeys, 10000);
    const monkeyBusinessLevel = calculateMonkeyBusiness(monkeys);
    console.log(monkeyBusinessLevel);
}

main();

File: elixir/day11_part2_2022.exs

defmodule MonkeyInMiddle do
  def solve do
    monkeys = parse_input("input.txt")
    run_rounds(monkeys, 10000, true)
  end

  def parse_input(path) do
    path
    |> File.read!()
    |> String.split("\n\n", trim: true)
    |> Enum.map(&parse_monkey/1)
    |> Enum.with_index()
    |> Enum.reduce(%{}, fn {monkey, index}, acc -> Map.put(acc, index, monkey) end)
  end

  def parse_monkey(monkey_desc) do
    lines = String.split(monkey_desc, "\n")
    
    items = 
      lines 
      |> Enum.at(1) 
      |> String.split(": ") 
      |> List.last() 
      |> String.split(", ") 
      |> Enum.map(&String.to_integer/1)

    [op1, operator, op2] = 
      lines 
      |> Enum.at(2) 
      |> String.split("= ") 
      |> List.last() 
      |> String.split()

    operation = create_operation(operator, op1, op2)
    
    divisible_by = 
      lines 
      |> Enum.at(3) 
      |> String.split() 
      |> List.last() 
      |> String.to_integer()
    
    true_target = 
      lines 
      |> Enum.at(4) 
      |> String.split() 
      |> List.last() 
      |> String.to_integer()
    
    false_target = 
      lines 
      |> Enum.at(5) 
      |> String.split() 
      |> List.last() 
      |> String.to_integer()

    %{
      items: items,
      operation: operation,
      divisible_by: divisible_by,
      targets: {true_target, false_target},
      inspections: 0
    }
  end

  defp create_operation("+", "old", "old"), do: fn x -> x + x end
  defp create_operation("+", "old", n), do: fn x -> x + String.to_integer(n) end
  defp create_operation("*", "old", "old"), do: fn x -> x * x end
  defp create_operation("*", "old", n), do: fn x -> x * String.to_integer(n) end

  def run_rounds(monkeys, rounds, worry_management) do
    divisor = calculate_divisor(monkeys)
    
    final_monkeys = 
      Enum.reduce(1..rounds, monkeys, fn _, acc -> 
        process_round(acc, divisor, worry_management) 
      end)

    final_monkeys
    |> Map.values()
    |> Enum.map(& &1.inspections)
    |> Enum.sort(:desc)
    |> Enum.take(2)
    |> Enum.product()
  end

  defp calculate_divisor(monkeys) do
    monkeys
    |> Map.values()
    |> Enum.map(& &1.divisible_by)
    |> Enum.product()
  end

  defp process_round(monkeys, divisor, worry_management) do
    Enum.reduce(0..(map_size(monkeys) - 1), monkeys, fn monkey_index, acc ->
      process_monkey(acc, monkey_index, divisor, worry_management)
    end)
  end

  defp process_monkey(monkeys, monkey_index, divisor, worry_management) do
    monkey = Map.get(monkeys, monkey_index)
    
    Enum.reduce(monkey.items, monkeys, fn item, acc ->
      new_worry = 
        monkey.operation.(item)
        |> then(fn w -> 
          if worry_management, do: rem(w, divisor), else: div(w, 3) 
        end)
      
      target = 
        if rem(new_worry, monkey.divisible_by) == 0, 
          do: elem(monkey.targets, 0), 
          else: elem(monkey.targets, 1)
      
      acc
      |> update_in([monkey_index, :items], &List.delete(&1, item))
      |> update_in([monkey_index, :inspections], &(&1 + 1))
      |> update_in([target, :items], fn items -> items ++ [new_worry] end)
    end)
  end
end

# Run the solution
IO.puts(MonkeyInMiddle.solve())
